File: Python3/170. Two Sum III - Data structure design.py
class TwoSum:

    def __init__(self):
        self.nums = collections.defaultdict(int)

    def add(self, number: int) -> None:
        self.nums[number] += 1
        
    def find(self, value: int) -> bool:
        for num, idx in self.nums.items():
            if value-num in self.nums:
                # if value-num == num, there must be 2 or more num
                if value-num == num:
                    if self.nums[num] >= 2:
                        return True
                    
                else:
                    return True
                
        return False


# Your TwoSum object will be instantiated and called as such:
# obj = TwoSum()
# obj.add(number)
# param_2 = obj.find(value)

File: Java/170. Two Sum III - Data structure design.java
class TwoSum {
    
    Map<Integer, Integer> nums;
    
    public TwoSum() {
        nums = new HashMap<>();
    }
    
    public void add(int number) {
        nums.put(number, nums.getOrDefault(number, 0)+1);
    }
    
    public boolean find(int value) {
        for(Integer num: nums.keySet()){
            if(nums.containsKey(value-num)){
                if(value-num == num){
                    if(nums.get(num) >= 2){
                        return true;
                    }
                }else{
                    return true;
                }
            }
        }
        return false;
    }
}

/**
 * Your TwoSum object will be instantiated and called as such:
 * TwoSum obj = new TwoSum();
 * obj.add(number);
 * boolean param_2 = obj.find(value);
 */